### PR TITLE
ref(notifications): Move Activities to Notifications

### PR DESCRIPTION
**Pull request by mgaeta**
The workflow notification classes currently live in the mail module. This PR renames and moves them to the notifications module so that they can be used with other notification types besides email.

### CODE REVIEW

File: src/sentry/notifications/activity/unassigned.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+from .base import ActivityNotification
+
+
+class UnassignedActivityNotification(ActivityNotification):
+    def get_activity_name(self) -> str:
+        return "Unassigned"
+
+    def get_description(self) -> Tuple[str, Mapping[str, Any], Mapping[str, Any]]:
+        return "{author} unassigned {an issue}", {}, {}
```
Comment (ceorourke) [inline src/sentry/notifications/activity/unassigned.py:11]: can we use an f string here?

File: src/sentry/models/activity.py [hunk 2]
```diff
@@ -11,67 +11,43 @@
...
+    SET_UNRESOLVED = ActivityType.SET_UNRESOLVED.value
+    UNASSIGNED = ActivityType.UNASSIGNED.value
+    UNMERGE_DESTINATION = ActivityType.UNMERGE_DESTINATION.value
+    UNMERGE_SOURCE = ActivityType.UNMERGE_SOURCE.value
 
     project = FlexibleForeignKey("sentry.Project")
     group = FlexibleForeignKey("sentry.Group", null=True)
     # index on (type, ident)
-    type = BoundedPositiveIntegerField(choices=TYPE)
+    type = BoundedPositiveIntegerField(
```
Comment (ceorourke) [inline src/sentry/models/activity.py:48]: this change is causing us to need a migration, is it totally necessary?

File: src/sentry/notifications/activity/base.py [hunk 3]
```diff
@@ -15,30 +16,30 @@
...
         self.organization = self.project.organization
         self.group = activity.group
 
-    def _get_subject_prefix(self):
+    def _get_subject_prefix(self) -> str:
         prefix = ProjectOption.objects.get_value(project=self.project, key="mail:subject_prefix")
         if not prefix:
             prefix = options.get("mail.subject-prefix")
-        return prefix
+        return str(prefix)
```
Comment (ceorourke) [inline src/sentry/notifications/activity/base.py:30]: why does this need to be converted to a string now?

File: src/sentry/notifications/activity/base.py [hunk 4]
```diff
@@ -15,30 +16,30 @@
...
-    def should_email(self):
+    def should_email(self) -> bool:
         return True
 
-    def get_participants(self):
+    def get_participants(self) -> Mapping[User, int]:
         # TODO(dcramer): not used yet today except by Release's
         if not self.group:
-            return []
+            return {}
```
Comment (ceorourke) [inline src/sentry/notifications/activity/base.py:38]: So this used to be a list of users, and now it's a mapping of users to what? Why is it changing in this PR?

File: src/sentry/notifications/activity/base.py [hunk 5]
```diff
@@ -177,9 +173,9 @@ def _get_user_avatar_url(self, user, size=20):
         url = reverse("sentry-user-avatar-url", args=[avatar.ident])
         if size:
             url = f"{url}?s={int(size)}"
-        return absolute_uri(url)
+        return str(absolute_uri(url))
```
Comment (ceorourke) [inline src/sentry/notifications/activity/base.py:176]: same here, why is this a string now? is it just because of the typing?

File: src/sentry/notifications/activity/release.py [hunk 6]
```diff
@@ -157,29 +158,23 @@ def get_participants(self):
                 or NotificationSettingOptionValues.COMMITTED_ONLY.value  # product default
             )
 
-        # filter down to members which have been seen in the commit log:
-        participants_committed = {
-            user: GroupSubscriptionReason.committed
-            for user, option in users_with_options.items()
-            if (
+        users_to_reasons: MutableMapping[User, int] = {}
```
Comment (ceorourke) [inline src/sentry/notifications/activity/release.py:161]: does this logic change belong in this PR? seems unrelated to moving files around

File: src/sentry/notifications/activity/unassigned.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/notifications/activity/unassigned.py:11]: This string is passed as `description` to: 
```
return description.format(**context)
```
so it needs to remain is this weird format.

File: src/sentry/notifications/activity/base.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/notifications/activity/base.py:31]: `get_value` and `options.get` return `Any`. We know its a string so I'm wrapping it in `str` for the type checker.

File: src/sentry/notifications/activity/base.py (hunk 4 above)
Comment (mgaeta) [reply src/sentry/notifications/activity/base.py:39]: This was probably a bug. The expected return type of `get_participants` is changing from `Mapping[User, int]` to `Mapping[ExternalProviders, Mapping[User, int]]`.

File: src/sentry/notifications/activity/base.py (hunk 5 above)
Comment (mgaeta) [reply src/sentry/notifications/activity/base.py:176]: Yeah it's for types. I marked `absolute_uri` as returning an `str` but `mypy` doesn't believe me.

File: src/sentry/notifications/activity/release.py (hunk 6 above)
Comment (mgaeta) [reply src/sentry/notifications/activity/release.py:161]: The logic isn't changing but I had to the refactor because the type checker doesn't like  `itertools.chain`.

File: src/sentry/types/activity.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    # The user has reprocessed the group, so events may have moved to new groups
+    REPROCESS = 22
+    MARK_REVIEWED = 23
+
+
+# Warning: This must remain in this EXACT order.
+CHOICES = tuple(
+    (i, ActivityType(i).name.lower())
+    for i in [1, 15, 13, 16, 21, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 17, 18, 19, 20, 22, 23]
```
Comment (ceorourke) [inline src/sentry/types/activity.py:35]: I'm worried about this, is it really so bad to copy the `TYPE` from the other file instead?

Comment (mgaeta) [reply src/sentry/types/activity.py:35]: I think its worth it. The alternative gets pretty unreadable because of the enums:
```
SLUGS = {
    ActivityType.ASSIGNED: "assigned",
    ActivityType.CREATE_ISSUE: "create_issue",
    ActivityType.DEPLOY: "deploy",
    ActivityType.FIRST_SEEN: "first_seen",
    ActivityType.MARK_REVIEWED: "mark_reviewed",
    ActivityType.MERGE: "merge",
    ActivityType.NEW_PROCESSING_ISSUES: "new_processing_issues",
    ActivityType.NOTE: "note",
    ActivityType.RELEASE: "release",
    ActivityType.REPROCESS: "reprocess",
    ActivityType.SET_IGNORED: "set_ignored",
    ActivityType.SET_PRIVATE: "set_private",
    ActivityType.SET_PUBLIC: "set_public",
    ActivityType.SET_REGRESSION: "set_regression",
    ActivityType.SET_RESOLVED: "set_resolved",
    ActivityType.SET_RESOLVED_BY_AGE: "set_resolved_by_age",
    ActivityType.SET_RESOLVED_IN_COMMIT: "set_resolved_in_commit",
    ActivityType.SET_RESOLVED_IN_PULL_REQUEST: "set_resolved_in_pull_request",
    ActivityType.SET_RESOLVED_IN_RELEASE: "set_resolved_in_release",
    ActivityType.SET_UNRESOLVED: "set_unresolved",
    ActivityType.UNASSIGNED: "unassigned",
    ActivityType.UNMERGE_DESTINATION: "unmerge_destination",
    ActivityType.UNMERGE_SOURCE: "unmerge_source",
}

TYPE = (
    (ActivityType.SET_RESOLVED.value, SLUGS[ActivityType.SET_RESOLVED]),
    (ActivityType.SET_RESOLVED_BY_AGE.value, SLUGS[ActivityType.SET_RESOLVED_BY_AGE]),
    (ActivityType.SET_RESOLVED_IN_RELEASE.value, SLUGS[ActivityType.SET_RESOLVED_IN_RELEASE]),
    (ActivityType.SET_RESOLVED_IN_COMMIT.value, SLUGS[ActivityType.SET_RESOLVED_IN_COMMIT]),
    (ActivityType.SET_RESOLVED_IN_PULL_REQUEST.value, SLUGS[ActivityType.SET_RESOLVED_IN_PULL_REQUEST]),
    (ActivityType.SET_UNRESOLVED.value, SLUGS[ActivityType.SET_UNRESOLVED]),
    (ActivityType.SET_IGNORED.value, SLUGS[ActivityType.SET_IGNORED]),
    (ActivityType.SET_PUBLIC.value, SLUGS[ActivityType.SET_PUBLIC]),
    (ActivityType.SET_PRIVATE.value, SLUGS[ActivityType.SET_PRIVATE]),
    (ActivityType.SET_REGRESSION.value, SLUGS[ActivityType.SET_REGRESSION]),
    (ActivityType.CREATE_ISSUE.value, SLUGS[ActivityType.CREATE_ISSUE]),
    (ActivityType.NOTE.value, SLUGS[ActivityType.NOTE]),
    (ActivityType.FIRST_SEEN.value, SLUGS[ActivityType.FIRST_SEEN]),
    (ActivityType.RELEASE.value, SLUGS[ActivityType.RELEASE]),
    (ActivityType.ASSIGNED.value, SLUGS[ActivityType.ASSIGNED]),
    (ActivityType.UNASSIGNED.value, SLUGS[ActivityType.UNASSIGNED]),
    (ActivityType.MERGE.value, SLUGS[ActivityType.MERGE]),
    (ActivityType.DEPLOY.value, SLUGS[ActivityType.DEPLOY]),
    (ActivityType.NEW_PROCESSING_ISSUES.value, SLUGS[ActivityType.NEW_PROCESSING_ISSUES]),
    (ActivityType.UNMERGE_SOURCE.value, SLUGS[ActivityType.UNMERGE_SOURCE]),
    (ActivityType.UNMERGE_DESTINATION.value, SLUGS[ActivityType.UNMERGE_DESTINATION]),
    (ActivityType.REPROCESS.value, SLUGS[ActivityType.REPROCESS]),
    (ActivityType.MARK_REVIEWED.value, SLUGS[ActivityType.MARK_REVIEWED]),
)
```

Comment (ceorourke) [reply src/sentry/types/activity.py:35]: hmmm personally I prefer the long winded way (maybe a comment with the value would help?) but I'd be curious to see what others think here.

**Review by ceorourke — APPROVED** [review]
